File: frontend/src/lib/websockets.ts
import { LogEntry } from "@/components/logs/columns";
import { UserEntry } from "@/components/users/editForm/formComponents";
import { toast } from "sonner";

export const handleMessage = (message: WebsocketMessageBody) => {

    switch (message.action) {
        case MessageAction.CardRead:
            {
                toast.info("Read card: "+ message.data.card_serial_number + ".",
                    {
                        closeButton: true
                    })
                break;
            }
        case MessageAction.NewLogEntry:
            {

                toast.success("New log of user: " + message.data.email + ", with card_serial_number: " + message.data.card_serial_number + ".",
                    {
                        closeButton: true
                    })
                break;
            }
        case MessageAction.NewUserEntry:
            {
                break;
            }
    }
}

export enum MessageAction {
    CardRead = "CardRead",
    NewLogEntry = "NewLogEntry",
    NewUserEntry = "NewUserEntry"
}

export type WebsocketMessageBody = {
    action: MessageAction,
    data: LogEntry | UserEntry
}

File: backend/common/src/types/websockets.rs
use serde::Serialize;

use super::database::{CardRead, LogEntry, UserEntry};

#[derive(Serialize, Debug, Clone)]
pub enum MessageAction {
    CardRead,
    NewLogEntry,
    NewUserEntry,
}

#[derive(Serialize, Debug, Clone)]
#[serde(untagged)]
pub enum WebsocketMessageData {
    UserEntry(UserEntry),
    LogEntry(LogEntry),
    CardRead(CardRead)
}

#[derive(Serialize, sqlx::FromRow, Debug, Clone)]
pub struct WebsocketMessageBody {
    pub action: MessageAction,
    pub data: WebsocketMessageData,
}
